File: statsig-ffi/bindings/java/src/main/java/com/statsig/StatsigUser.java
package com.statsig;

import com.google.gson.Gson;

import java.util.Map;

public class StatsigUser {
    public String userID;
    public Map<String, String> customIDs;
    public String email;
    public String ip;
    public String locale;
    public String appVersion;
    public String userAgent;
    public String country;
    public Map<String, String> privateAttributes;
    public Map<String, Object> custom;

    private volatile String ref;

    public StatsigUser(
            String userId,
            Map<String, String> customIDs,
            String email,
            String ip,
            String userAgent,
            String country,
            String locale,
            String appVersion,
            Map<String, Object> custom,
            Map<String, String> privateAttributes
    ) {
        this.userID = userId;
        this.custom = custom;
        this.email = email;
        this.ip = ip;
        this.locale = locale;
        this.appVersion = appVersion;
        this.customIDs = customIDs;
        this.privateAttributes = privateAttributes;
        this.country = country;
        this.userAgent = userAgent;

        initializeRef();
        ResourceCleaner.register(this, () -> {
            if (ref != null) {
                StatsigJNI.statsigUserRelease(ref);
                ref = null;
            }
        });
    }

    public StatsigUser(String userId) {
        this(userId, null, null, null, null, null, null, null, null, null);
    }

    public StatsigUser(Map<String, String> customIDs) {
        this(null, customIDs, null, null, null, null, null, null, null, null);
    }

    public StatsigUser(String userId, String email) {
        this(userId, null, email, null, null, null, null, null, null, null);
    }

    public String getUserID() {
        return userID;
    }

    public void setUserID(String userID) {
        this.userID = userID;
    }

    public Map<String, String> getCustomIDs() {
        return customIDs;
    }

    public void setCustomIDs(Map<String, String> customIDs) {
        this.customIDs = customIDs;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getIp() {
        return ip;
    }

    public void setIp(String ip) {
        this.ip = ip;
    }

    public String getLocale() {
        return locale;
    }

    public void setLocale(String locale) {
        this.locale = locale;
    }

    public String getAppVersion() {
        return appVersion;
    }

    public void setAppVersion(String appVersion) {
        this.appVersion = appVersion;
    }

    public String getUserAgent() {
        return userAgent;
    }

    public void setUserAgent(String userAgent) {
        this.userAgent = userAgent;
    }

    public String getCountry() {
        return country;
    }

    public void setCountry(String country) {
        this.country = country;
    }

    public Map<String, String> getPrivateAttributes() {
        return privateAttributes;
    }

    public void setPrivateAttributes(Map<String, String> privateAttributes) {
        this.privateAttributes = privateAttributes;
    }

    public Map<String, Object> getCustom() {
        return custom;
    }

    public void setCustom(Map<String, Object> custom) {
        this.custom = custom;
    }

    private void initializeRef() {
        Gson gson = new Gson();

        String customIdsJson = customIDs != null ? gson.toJson(customIDs) : null;
        String customJson = custom != null ? gson.toJson(custom) : null;
        String privateAttributesJson = privateAttributes != null ? gson.toJson(privateAttributes) : null;

        // Pass all arguments to the JNI binding
        this.ref = StatsigJNI.statsigUserCreate(
                userID,
                customIdsJson,
                email,
                ip,
                userAgent,
                country,
                locale,
                appVersion,
                customJson,
                privateAttributesJson
        );
    }

    String getRef() {
        return ref;
    }
}

File: statsig-rust/src/user/statsig_user.rs
use crate::evaluation::dynamic_value::DynamicValue;
use crate::statsig_metadata;
use crate::{dyn_value, evaluation::dynamic_string::DynamicString};
use indexmap::IndexMap;
use std::sync::Arc;

use super::{
    into_optional::IntoOptional,
    unit_id::UnitID,
    user_data::{IntoOptionalUserDataMap, UserData, UserDataMap},
};

#[derive(Clone)]
pub struct StatsigUser {
    pub data: Arc<UserData>,

    pub(crate) sdk_version: &'static str,
}

impl StatsigUser {
    #[must_use]
    pub fn with_user_id(user_id: impl Into<UnitID>) -> Self {
        let unit_id: UnitID = user_id.into();
        Self::new(UserData {
            user_id: Some(unit_id.into()),
            ..UserData::default()
        })
    }

    #[must_use]
    pub fn with_custom_ids<K, U, I>(custom_ids: I) -> Self
    where
        I: IntoIterator<Item = (K, U)>,
        K: Into<String>,
        U: Into<UnitID>,
    {
        let custom_ids: UserDataMap = custom_ids
            .into_iter()
            .map(|(k, v)| (k.into(), v.into().into()))
            .collect();

        Self::new(UserData {
            custom_ids: Some(custom_ids),
            ..UserData::default()
        })
    }

    pub fn new(inner: UserData) -> Self {
        Self {
            data: Arc::new(inner),
            sdk_version: statsig_metadata::SDK_VERSION,
        }
    }
}

// -------------------------------------------------------------------------------- [Getters/Setters]

macro_rules! string_field_accessor {
    ($self:ident, $getter_name:ident, $setter_name:ident, $field:ident) => {
        pub fn $getter_name(&self) -> Option<&str> {
            self.data
                .$field
                .as_ref()?
                .string_value
                .as_ref()
                .map(|s| s.value.as_str())
        }

        pub fn $setter_name(&mut self, value: impl IntoOptional<String>) {
            let value = value.into_optional();
            let mut_data = Arc::make_mut(&mut self.data);
            match value {
                Some(value) => {
                    mut_data.$field = Some(dyn_value!(value));
                }
                None => mut_data.$field = None,
            }
        }
    };
}

macro_rules! map_field_accessor {
    ($self:ident, $getter_name:ident, $setter_name:ident, $field:ident) => {
        pub fn $getter_name(&self) -> Option<&UserDataMap> {
            self.data.$field.as_ref()
        }

        pub fn $setter_name(&mut self, value: impl IntoOptionalUserDataMap) {
            let mut_data = Arc::make_mut(&mut self.data);
            let value = match value.into_optional_user_data_map() {
                Some(value) => value,
                None => {
                    mut_data.$field = None;
                    return;
                }
            };

            mut_data.$field = Some(value);
        }
    };
}

impl StatsigUser {
    // ---------------------------------------- [User ID]

    pub fn get_user_id(&self) -> Option<&str> {
        self.data
            .user_id
            .as_ref()?
            .string_value
            .as_ref()
            .map(|s| s.value.as_str())
    }

    pub fn set_user_id(&mut self, user_id: impl Into<UnitID>) {
        let unit_id = user_id.into();
        let mut_data = Arc::make_mut(&mut self.data);
        mut_data.user_id = Some(unit_id.into());
    }

    // ---------------------------------------- [Custom IDs]

    pub fn get_custom_ids(&self) -> Option<IndexMap<&str, &str>> {
        let mapped = self
            .data
            .custom_ids
            .as_ref()?
            .iter()
            .map(entry_to_key_value_refs)
            .collect();

        Some(mapped)
    }

    pub fn set_custom_ids<K, U, I>(&mut self, custom_ids: I)
    where
        I: IntoIterator<Item = (K, U)>,
        K: Into<String>,
        U: Into<UnitID>,
    {
        let custom_ids = custom_ids
            .into_iter()
            .map(|(k, v)| (k.into(), v.into().into()))
            .collect();

        let mut_data = Arc::make_mut(&mut self.data);
        mut_data.custom_ids = Some(custom_ids);
    }

    // ---------------------------------------- [Unit ID]

    pub fn get_unit_id(&self, id_type: &DynamicString) -> Option<&DynamicValue> {
        if id_type.lowercased_value.eq("userid") {
            return self.data.user_id.as_ref();
        }

        let custom_ids = self.data.custom_ids.as_ref()?;

        if let Some(custom_id) = custom_ids.get(id_type.value.as_str()) {
            return Some(custom_id);
        }

        custom_ids.get(id_type.lowercased_value.as_str())
    }

    // ---------------------------------------- [ Statsig Environment ]

    pub fn get_statsig_environment(&self) -> Option<IndexMap<&str, &str>> {
        let mapped = self
            .data
            .statsig_environment
            .as_ref()?
            .iter()
            .map(entry_to_key_value_refs)
            .collect();

        Some(mapped)
    }

    pub fn set_statsig_environment<K, U, I>(&mut self, statsig_environment: Option<I>)
    where
        I: IntoIterator<Item = (K, U)>,
        K: Into<String>,
        U: Into<String>,
    {
        let mut_data = Arc::make_mut(&mut self.data);

        let statsig_environment = match statsig_environment {
            Some(v) => v,
            None => {
                mut_data.statsig_environment = None;
                return;
            }
        };

        let statsig_environment: UserDataMap = statsig_environment
            .into_iter()
            .map(|(k, v)| (k.into(), v.into().into()))
            .collect();

        mut_data.statsig_environment = Some(statsig_environment);
    }

    // ---------------------------------------- [ String Fields ]

    string_field_accessor!(self, get_email, set_email, email);
    string_field_accessor!(self, get_ip, set_ip, ip);
    string_field_accessor!(self, get_user_agent, set_user_agent, user_agent);
    string_field_accessor!(self, get_country, set_country, country);
    string_field_accessor!(self, get_locale, set_locale, locale);
    string_field_accessor!(self, get_app_version, set_app_version, app_version);

    // ---------------------------------------- [ Map Fields ]

    map_field_accessor!(self, get_custom, set_custom, custom);
    map_field_accessor!(
        self,
        get_private_attributes,
        set_private_attributes,
        private_attributes
    );
}

// -------------------------------------------------------------------------------- [Helpers]

fn entry_to_key_value_refs<'a>(entry: (&'a String, &'a DynamicValue)) -> (&'a str, &'a str) {
    let (key, value) = entry;

    (
        key.as_str(),
        value
            .string_value
            .as_ref()
            .map(|s| s.value.as_str())
            .unwrap_or(""),
    )
}
